File: leetcode/P200.java
class Solution {
  private boolean[][] visited;

  public void dfs(char[][] grid, int i, int j) {
    if (i >= grid.length || j >= grid[0].length || i < 0 || j < 0 || visited[i][j] == true || grid[i][j] == '0') {
      return;
    }

    visited[i][j] = true;
    dfs(grid, i + 1, j);
    dfs(grid, i, j + 1);
    dfs(grid, i - 1, j);
    dfs(grid, i, j - 1);
  }

  public int numIslands(char[][] grid) {
    if (grid == null || grid.length == 0) {
      return 0;
    }
    visited = new boolean[grid.length][grid[0].length];
    int count = 0;
    for (int i = 0; i < grid.length; i++) {
      for (int j = 0; j < grid[0].length; j++) {
        if (grid[i][j] == '1' && visited[i][j] == false) {
          dfs(grid, i, j);
          count++;
        }
      }
    }
    return count;
  }
}

File: leetcode/P200.py
def dfs(grid, x, y):
    if (0 <= x < len(grid)) and (0 <= y < len(grid[0])) and grid[x][y] == "1":
        grid[x][y] = -1
        dirs = [[0, 1], [1, 0], [0, -1], [-1, 0]]
        for dir in dirs:
            newX = dir[0] + x
            newY = dir[1] + y
            dfs(grid, newX, newY)


class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        count = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == "1":
                    dfs(grid, i, j)
                    count += 1
        return count
